### ISSUE TITLE
Implimenting k-opt

### ISSUE THREAD

**Issue by Stevo15025** [Stevo15025/tsp_gpu#21]
I don't think it would be that large of an extension to include a k-opt swap and/or insert kernel

https://pdfs.semanticscholar.org/12a4/5bd52921b1b86d610976b25e12f63bc96d22.pdf

**Comment by PB12203006** [comment]
It's executable, and I think after implementing this, we actually need no more annealing...Because I believe three different exchange rules can overcome most local minimum, i.e. saddle points in energy space, and get the global one directly....

**Comment by PB12203006** [comment]
I've finished it. Since the article said the high dimensional opt can be decomposed to 2_Opt I only implemented 2.

I've tried it on mona-lisa and the result is satisfying, the loss monotonically decreases!! And there's nothing wrong with the output route!

But because of the unsolved error my code is still a previous version, so I'll pull an independent file containing the 2_opt, basically it's similar to the other two, you can understand it!

**Comment by PB12203006** [comment]
![image](https://cloud.githubusercontent.com/assets/23200814/20915120/df82f4fe-bb51-11e6-955a-6c0dd18e98ed.png)

look, in update part we actually should revers the red sequence, from i+1 to j

**Comment by PB12203006** [comment]
Remember to adjust the sample space according to your T range

**Comment by Stevo15025** [comment]
Am I correct in thinking, in the linear form, the above graph is 

Old Trip:
0**123**456**789**0
New Trip:
0**789**456**123**0

**Comment by PB12203006** [comment]
No, 2 opt only changes two segments, but although the length of the red part remains, the direction has been reversed.

For example, if I choose segment (2,3) and (7,8)

01**23**456**78**90
to
01**27**654**38**90
"456" becomes "654" but (4,5)+(5,6)=(6,5)+(5,4)

**Comment by PB12203006** [comment]
then, (2,3) and (7,8) is two green segments in the picture, and they becomes(2,7) and (3,8)

but when implementing it's easy, just reverse the sequence 34567 to 76543, no edge condition

**Comment by PB12203006** [comment]
And notice that in this algo, we must guarantee i+1<j, to make sure there's two previous segments and two post segments

**Comment by Stevo15025** [comment]
Did you bold the wrong numbers there?

012**34**5**67**890
to
012**76**5**43**890

**Comment by PB12203006** [comment]
nono, note that (i,j) is equivalent to (j,i)
so, in route 01276543890 the real change of segments is (2,3) and (7,8), they becomes (2,7) and (3,8),
the (3,4)(4,5)(5,6)(6,7) all remain the same between the change, they just changed the direction and you can not figure them out when just connect the points with a line without arrow
(3,4)(4,5)(5,6)(6,7) to (7,6)(6,5)(5,4)(4,3), euivalent

**Comment by PB12203006** [comment]
but 2 and 8 are fixed, it's natural, because we change the other dot of their segments. changing the segment only needs to change one point of it.

**Comment by PB12203006** [comment]
large k-opt has many ways to link the points, that's the main problem, maybe you can try the sequential way to link, that's easy to implement by loop.
![image](https://cloud.githubusercontent.com/assets/23200814/20917039/2c1739c8-bb5d-11e6-8964-730b5c61fd54.png)

**Comment by PB12203006** [comment]
Note only the arrow contains start point moves clockwise

start and end is special, edge condition
link (i_1), (i_2)    reverse (i_1)+1...i_2

this part can be iterated
link (i_1)+1, (i_3)    reverse (i_2)+1...i_3
link (i_2)+1, (i_4)    reverse (i_3)+1...i_4
....

end edge condition
link i_(end-1)+1, (i_end)+1, no reverse

**Comment by PB12203006** [comment]
Got a flu shot today, I'm tired, not much sober, made some grammar mistake. Forgive me, going to sleep now!

intuitively I don't think large k-opt has some incomparable advantage...Perhaps I'm not sober now...

**Comment by Stevo15025** [comment]
Haha okay very reasonable! I think I'm getting the picture.
